src/OpenTelemetry.Instrumentation.Wcf/Implementation/RequestTelemetryStateTracker.cs

```diff
@@ -0,0 +1,162 @@
+    private Timer timer;
+
+    private long configuredTimeoutInMs;
+
+    public RequestTelemetryStateTracker(TimeSpan timeout)
+    {
+        this.configuredTimeoutInMs = (long)timeout.TotalMilliseconds;
+    }
+
+    public event EventHandler<RequestTelemetryState> TelemetryStateTimedOut;
+
+    public void PushTelemetryState(Message request, RequestTelemetryState telemetryState)
+    {
+        var messageId = request?.Headers.MessageId?.ToString();
+        if (messageId != null)
+        {
+            var expiresAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + this.configuredTimeoutInMs;
+            var entry = new Entry(telemetryState, expiresAt);
+            lock (this.sync)
+            {
+                this.outstandingRequestStates.Add(messageId, entry);
+                if (this.timer == null)
+                {
+                    this.SetTimer(expiresAt);
+                }
+            }
+        }
+    }
+
+    public RequestTelemetryState PopTelemetryState(Message reply)
+    {
+        var relatesTo = reply?.Headers.RelatesTo?.ToString();
+        return relatesTo == null ? null : this.PopTelemetryState(relatesTo);
+    }
+
+    public void Dispose()
+    {
+        this.SetTimer(null);
+
+        Entry[] abandonedEntries = null;
+        lock (this.sync)
+        {
+            abandonedEntries = new Entry[this.outstandingRequestStates.Count];
+            this.outstandingRequestStates.Values.CopyTo(abandonedEntries, 0);
+            this.outstandingRequestStates.Clear();
+        }
+
+        foreach (Entry entry in abandonedEntries)
+        {
+            this.TriggerTimedOutEvent(entry);
+        }
+    }
+
+    private RequestTelemetryState PopTelemetryState(string messageId)
+    {
+        lock (this.sync)
+        {
+            var entry = (Entry)this.outstandingRequestStates[messageId];
+            if (entry != null)
+            {
+                this.outstandingRequestStates.Remove(messageId);
+            }
+
+            return entry?.State;
+        }
+    }
+
+    private void OnTimer(object state)
+    {
+        List<DictionaryEntry> timedOut = null;
+        lock (this.sync)
+        {
+            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            long? nextExpiry = null;
+            foreach (DictionaryEntry kvp in this.outstandingRequestStates)
+            {
+                if (((Entry)kvp.Value).ExpiresAt <= now)
+                {
+                    if (timedOut == null)
+                    {
+                        timedOut = new List<DictionaryEntry>();
+                    }
+
+                    timedOut.Add(kvp);
+                }
+                else
+                {
+                    nextExpiry = ((Entry)kvp.Value).ExpiresAt;
+                    break;
+                }
+            }
+
+            timedOut?.ForEach(kvp => this.outstandingRequestStates.Remove(kvp.Key));
+            this.SetTimer(nextExpiry);
```

Comment:
It's non-obvious that if there are no non-expiring items in outstandingRequestStates that nextExpiry will be null which will in effect cancel the timer. It might be worth adding a comment explaining this.